driver-core/src/test/unit/com/mongodb/internal/connection/DefaultServerMonitorTest.java

```diff
@@ -0,0 +1,300 @@
+
+        // When
+        TestServerMonitorListener listener = createTestServerMonitorListener();
+        monitor = createAndStartMonitor(createConnectionFactory(mockConnection), listener);
+
+        listener.waitForEvents(ServerHeartbeatFailedEvent.class, event -> true, 1, Duration.ofSeconds(30));
+        ServerHeartbeatStartedEvent startedEvent = getEvent(ServerHeartbeatStartedEvent.class, listener);
+        ServerHeartbeatFailedEvent failedEvent = getEvent(ServerHeartbeatFailedEvent.class, listener);
+
+        // Then
+        assertEquals(connectionDescription.getConnectionId(), startedEvent.getConnectionId());
+        assertEquals(connectionDescription.getConnectionId(), failedEvent.getConnectionId());
+        assertEquals(exception, failedEvent.getThrowable());
+        assertTrue(failedEvent.getElapsedTime(TimeUnit.NANOSECONDS) > 0);
+    }
+
+    @Test
+    void shouldEmitHeartbeatStartedBeforeSocketIsConnected() throws Exception {
+        // Given
+        InternalConnection mockConnection = mock(InternalConnection.class);
+        CountDownLatch latch = new CountDownLatch(1);
+        List<String> events = new ArrayList<>();
+        ServerMonitorListener listener = new ServerMonitorListener() {
+                @Override
+                public void serverHearbeatStarted(final ServerHeartbeatStartedEvent event) {
+                    events.add("serverHeartbeatStartedEvent");
+                }
+
+                @Override
+                public void serverHeartbeatSucceeded(final ServerHeartbeatSucceededEvent event) {
+                    events.add("serverHeartbeatSucceededEvent");
+                    latch.countDown();
+                }
+
+                @Override
+                public void serverHeartbeatFailed(final ServerHeartbeatFailedEvent event) {
+                    events.add("serverHeartbeatFailedEvent");
+                    latch.countDown();
+                }
+            };
+
+        doAnswer(invocation -> {
+            events.add("client connected");
+            return null;
+        }).when(mockConnection).open(any());
+
+        when(mockConnection.getBuffer(anyInt())).thenReturn(new ByteBufNIO(ByteBuffer.allocate(1024)));
+        when(mockConnection.getDescription()).thenReturn(createDefaultConnectionDescription());
+        when(mockConnection.getInitialServerDescription()).thenReturn(createDefaultServerDescription());
+
+        doAnswer(invocation -> {
+            events.add("client hello received");
+            throw new SocketException("Socket error");
+        }).when(mockConnection).receive(any(), any());
+
+        // When
+        monitor = createAndStartMonitor(createConnectionFactory(mockConnection), listener);
+        assertTrue(latch.await(5, TimeUnit.SECONDS), "Timed out waiting for heartbeat");
+
+        // Then
+        List<String> expectedEvents = asList("serverHeartbeatStartedEvent", "client connected", "client hello received", "serverHeartbeatFailedEvent");
```

Comment:
This test method corresponds to https://github.com/mongodb/specifications/pull/1483/files#diff-af358d9f2fc605678592489c97e6d557b5fa0fcb447f6619d503dc7a62c1ee71R268

But what do the other methods correspond to? Should this method be part of a `...ProseTest.java` class?

Reply:
Good shout. I've added an (ignored) method in the `ServerDiscoveryAndMonitoringProseTests` to point to this test. As it requires mocking and package access to `com.mongodb.internal.connection` I couldn't add it there.